kv/dist_sender.go

```diff
@@ -562,57 +548,56 @@ func (ds *DistSender) Send(call client.Call) {
 						return util.RetryContinue, nil
 					}
 				}
-			} else if isMulti {
-				// If this request spans ranges, collect the replies.
-				responses = append(responses, reply)
-
-				// If this request has a bound, such as MaxResults in ScanRequest,
-				// check whether enough rows are got in this round.
-				if args, ok := args.(proto.Bounded); ok && args.GetBound() > 0 {
-					bound := args.GetBound()
-					if reply, ok := reply.(proto.Countable); ok {
-						if nextBound := bound - reply.Count(); nextBound > 0 {
-							// Update bound for the next round.
-							args.SetBound(nextBound)
-						} else {
-							// Set flag to break the loop.
-							descNext = nil
-						}
-					}
-				}
+				return util.RetryBreak, err
+			}
 
-				// descNext can be nil in two cases:
-				// 1. Got enough rows in the middle of the request.
-				// 2. It is the last range of the request.
-				if descNext == nil {
-					// Combine multiple responses into the first one.
-					for _, r := range responses[1:] {
-						// We've already ascertained earlier that we're dealing with a
-						// Combinable response type.
-						responses[0].(proto.Combinable).Combine(r)
+			// If this request has a bound, such as MaxResults in
+			// ScanRequest, check whether enough rows have been retrieved.
+			if args, ok := args.(proto.Bounded); ok && args.GetBound() > 0 {
+				bound := args.GetBound()
+				if reply, ok := reply.(proto.Countable); ok {
+					if nextBound := bound - reply.Count(); nextBound > 0 {
+						// Update bound for the next round.
+						args.SetBound(nextBound)
+					} else {
+						// Set flag to break the loop.
+						descNext = nil
 					}
-
-					// Write the final response back.
-					gogoproto.Merge(call.Reply, responses[0])
 				}
 			}
+
+			if call.Reply != reply {
+				// This is a multi-range request. Combine the new response
+				// with the existing response.
+				call.Reply.(proto.Combinable).Combine(reply)
+			}
 			return util.RetryBreak, err
 		})
+
 		// Immediately return if querying a range failed non-retryably.
 		// For multi-range requests, we return the failing range's reply.
 		if err != nil {
-			reply.Header().SetGoError(err)
-			gogoproto.Merge(call.Reply, reply) // Only relevant in multi-range case.
+			call.Reply.Header().SetGoError(err)
+			// if call.Reply != reply {
+			// 	gogoproto.Merge(call.Reply, reply) // Only relevant in multi-range case.
+			// }
 			return
 		}
 		// If this was the last range accessed by this call, exit loop.
 		if descNext == nil {
 			break
 		}
+
 		// In next iteration, query next range.
 		args.Header().Key = descNext.StartKey
 		// "Untruncate" EndKey to original.
-		args.Header().EndKey = call.Args.Header().EndKey
+		args.Header().EndKey = header.EndKey
+
+		if reply == call.Reply {
```

Comment:
For single range requests, `reply == call.Reply` will always be true, but we would have already broken out of the loop because `descNext == nil`. For multi-range requests, `reply == call.Reply` will be true for the first range, but not subsequent ranges.